Refactor Timelion expression suggestion logic for clarity.

cjcenizal opened the pull request:
Extracts some work from https://github.com/elastic/kibana/pull/11164.

## Overview

The current version of this logic involves DOM manipulation from within a directive. We can simplify this by leveraging Angular's directives.

## Changes

- Create timelionExpressionInput directive.
- Create timelionExpressionSuggestions directive.
- Create expression_directive_helpers service (which can be unit tested).

rashidkpc commented:
This comment isn't a review of this pull, just a good place to leave some notes about suggestions.

Timelion's autocomplete has always been one of those "Well, its better than nothing" sort of things. There's a quick digest of how I'd like to see autocompletion work here: https://github.com/cjcenizal/kibana/blob/ed3862e0abb35a2e7e69355d6993a0af166de3ea/src/core_plugins/timelion/public/directives/expression_directive.js#L14-L37

This pull should help with that process as it breaks out the logic for finding and inserting into the helpers file. Just needs someone to sit down and sort it out. Its probably not that hard to implement, the AST already has all of the lookup objects to make determining context for caret position simple.

{
   "tree": [...],
   "functions": [
      {
         "type": "function",
         "function": "es",
         "arguments": [...],
         "location": {
            "min": 1,
            "max": 15
         },
         "text": ".es(index=bar)"
      }
   ],
   "args": [
      {
         "type": "namedArg",
         "name": "index",
         "value": {
            "type": "literal",
            "value": "bar",
            "location": {
               "min": 11,
               "max": 14
            },
            "text": "bar"
         },
         "location": {
            "min": 5,
            "max": 14
         },
         "text": "index=bar",
         "function": "es"
      }
   ],
   "variables": {}
}

 Basically we just need to take the caret position and compare it to the `location.min` and `location.max` in each object in `functions` and `args`. That should give you an idea of if the user is typing a new function, new argument or an argument value. Once you know that you could use one of 3 suggester method to know what to pass to `insertAtLocation`. The only method currently implemented is for functions, but it would be trivial to add the user 2 once the caret context thing is sorted.

rashidkpc commented:
This changes the behavior of the "other" selection in the interval chooser, it now opens a 2nd dropdown instead of making the existing dropdown appear to be editable

rashidkpc commented:
It looks like some other CSS changes at some point had partial broken the intended behavior as well. This used to just appear as an editable drop down, it looks like some CSS changed at some point that made it go wacky.
![before_other](https://cloud.githubusercontent.com/assets/1250387/25502128/3e9f0b86-2b4a-11e7-9ec6-43ab5b6e27f9.gif)

In any case, the behavior of this pull isn't correct.
![after_other](https://cloud.githubusercontent.com/assets/1250387/25502147/55870290-2b4a-11e7-9c6d-b0abab434411.gif)

cjcenizal commented:
@rashidkpc Thanks for catching that. Fixed:

![image](https://cloud.githubusercontent.com/assets/1238659/25508099/ce1ef0a4-2b64-11e7-8bf9-c166b8367eac.png)

![image](https://cloud.githubusercontent.com/assets/1238659/25508100/cfb0c4a6-2b64-11e7-842f-6fa38d59af53.png)

ppisljar commented:
![timelion](https://cloud.githubusercontent.com/assets/13629809/25516411/c6ed925a-2bea-11e7-8f38-d1f714a85a2a.png)

with other dropdown selection (and second field appearing) ... still doens't look completely OK ... you can see that the dropdown gets super narrow, hardly fitting the values, where before selecting `other` there is 'the right amount' of padding around the values.

ppisljar commented on src/core_plugins/timelion/public/index.html:
```diff
@@ -24,17 +24,14 @@
+          <timelion-interval
+            model="state.interval"
+            in-header="true"
+          ></timelion-interval>
```
should we `/>` as there is nothing inside the timelion-interval element ?

cjcenizal replied:
Angular requires you to actually use a closing tag for custom directives.

ppisljar commented on src/core_plugins/timelion/public/directives/interval/interval.js:
```diff
@@ -9,6 +9,7 @@ app.directive('timelionInterval', function ($compile, $timeout) {
     restrict: 'E',
     scope: {
       model: '=', // The interval model
+      inHeader: '=',
```
how/where is this used ?

cjcenizal replied:
This is used to differentiate between this UI in the context of the Timelion app and the Visualize app (when you create a Timelion vis). I can add a comment about this.

ppisljar commented on src/core_plugins/timelion/public/directives/interval/interval.html:
```diff
@@ -1,8 +1,13 @@
-<select class="form-control timelion-interval timelion-interval--select"
+<select
+  class="timelion-interval"
   ng-options="i for i in intervalOptions"
```
`i for i` ... you can probably drop one `i`

cjcenizal replied:
I think this is the syntax [ng-options](https://docs.angularjs.org/api/ng/directive/ngOptions) expects. Is there a different expression we should use?

ppisljar replied:
no you are right, maybe just rename to intervalOption ?

cjcenizal commented:
@ppisljar Thanks for the review! I addressed your comments. Would you mind check out 3e5352574fffd3a4368341df6c77402ff3005e86 and letting me know if the select bug exists in that commit? If so, I think we should mark it in an issue since it's pre-existing, and address it in another PR.

rashidkpc commented:
Here's what I was going for:

![interval2](https://cloud.githubusercontent.com/assets/1250387/25598360/6da76cd8-2e89-11e7-87f5-e2a6e7706260.gif)

Here's the CSS I used:

```
select.timelion-interval {
  width: 80px;
}

input.timelion-interval {
  width: 50px;
}

select.timelion-interval-compact {
  width: 10px;
  padding-left: 0;
  border-left: none;
}
```

cjcenizal commented:
Thanks @rashidkpc. I tweaked your suggestion a bit and landed on this:

![timelion_interval](https://cloud.githubusercontent.com/assets/1238659/25600298/ef0a7cb2-2e97-11e7-8a73-724d878efcc3.gif)

I don't want to remove the border from the select because it makes the focus state look funky, and also makes the hit area harder to see. Could you take another look?

rashidkpc commented:
That works for me

cjcenizal commented:
jenkins test this

ppisljar commented:
@cjcenizal the dropdown issue is not present in 3e53525  (it works ok there). it happens in latest chrome, looks better in firefox.

also another small issue i found: when you click the dropdown (in firefox) you can see the "focus border" on the input that you don't really see .... 

![timeion_dropdown](https://cloud.githubusercontent.com/assets/13629809/25651331/c8955f60-2fe2-11e7-99b0-673c7d54ca8a.png)
